src/tabs/presets/CliEngine.js

```diff
@@ -0,0 +1,253 @@
+                this.executeCommands(outString);
+                textarea.val('');
+            }
+        });
+
+        // give input element user focus
+        textarea.focus();
+    }
+
+    close(callback)
+    {
+        this.send(this.getCliCommand('exit\r', ""), function () { //this.cliBuffer
+            if (callback) {
+                callback();
+            }
+        });
+    }
+
+    executeCommands(outString) {
+        const outputArray = outString.split("\n");
+        Promise.reduce(outputArray, (delay, line, index) => {
+            return new Promise((resolve) => {
+                GUI.timeout_add('CLI_send_slowly', () => {
+                    let processingDelay = this.lineDelayMs;
+                    line = line.trim();
+
+                    if (line.toLowerCase().startsWith('profile')) {
+                        processingDelay = this.profileSwitchDelayMs;
+                    }
+
+                    const isLastCommand = outputArray.length === index + 1;
+
+                    if (isLastCommand && this.cliBuffer) {
+                        line = this.getCliCommand(line, this.cliBuffer);
+                    }
+
+                    this.sendLine(line, function () {
+                        resolve(processingDelay);
+                    });
+                }, delay);
+            });
+        }, 0);
+    }
+
+    removePromptHash(promptText)
+    {
+        return promptText.replace(/^# /, '');
+        //return promptText;
+    }
+
+    cliBufferCharsToDelete(command, buffer) {
+        let commonChars = 0;
+        for (let i = 0; i < buffer.length; i++) {
+            if (command[i] === buffer[i]) {
+                commonChars++;
+            } else {
+                break;
+            }
+        }
+        return buffer.length - commonChars;
+    }
+
+    commandWithBackSpaces(command, buffer, noOfCharsToDelete) {
+        const backspace = String.fromCharCode(127);
+        return backspace.repeat(noOfCharsToDelete) + command.substring(buffer.length - noOfCharsToDelete, command.length);
+    }
+
+    getCliCommand(command, cliBuffer) {
+        const buffer = this.removePromptHash(cliBuffer);
+        const bufferRegex = new RegExp('^' + buffer, 'g');
+
+        if (command.match(bufferRegex)) {
+            const result = command.replace(bufferRegex, '');
+            return result;
+        }
+
+        const noOfCharsToDelete = this.cliBufferCharsToDelete(command, buffer);
+        const result = this.commandWithBackSpaces(command, buffer, noOfCharsToDelete);
+        return result;
+    }
+
+
+    writeToOutput(text)
+    {
+        this._windowWrapper.append(text);
+        this._window.scrollTop(this._windowWrapper.height());
+    }
+
+    writeLineToOutput(text) {
+        if (text.startsWith("###ERROR")) {
+            this.writeToOutput(`<span class="error_message">${text}</span><br>`);
+        } else {
+            this.writeToOutput(text + "<br>");
+        }
+    }
+
+    readSerial(readInfo)
```

Comment:
This is also used in cli, can we make this into util module function?

Reply:
The whole CLI engine class theoretically should be reused for CLI tab... However CLI tab code is a huuge mess at least to my small brain. It took me good 5 days to separate the minimum functionality of CLI to here.

I hope someone could help to merge CLI with CLI engine and make necessary modifications for CLI Engine for that. Or it can be in the future improvements...

Reply:
Guess it's best to make a util file first, and do the merging of cli in another PR as this is rather a critical part of configurator and keep this PR clean.